File: meteor-client/src/main/java/meteor/plugins/agility/AgilityPlugin.java
package meteor.plugins.agility;

import meteor.Plugin;
import meteor.eventbus.Subscribe;
import net.runelite.api.*;
import net.runelite.api.Point;
import net.runelite.api.events.*;

import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AgilityPlugin extends Plugin {

    private final Map<TileObject, Obstacle> obstacles = new HashMap<>();
    public List<TileItem> marks = new ArrayList<>();

    Color overlayFillColor;
    private Color activeColor;
    private final int OFFSET_Z = 20;
    private final String MARK_OF_GRACE = "Mark of Grace";

    @Subscribe
    public void onGameStateChanged(GameStateChanged event)
    {
        if (event.getGameState() == GameState.LOADING) {
            marks.clear();
            obstacles.clear();
        }
    }

    @Subscribe
    public void onGameObjectSpawned(GameObjectSpawned event)
    {
        onTileObject(event.getTile(), null, event.getGameObject());
    }

    @Subscribe
    public void onGameObjectChanged(GameObjectChanged event)
    {
        onTileObject(event.getTile(), event.getPrevious(), event.getGameObject());
    }

    @Subscribe
    public void onGameObjectDespawned(GameObjectDespawned event)
    {
        onTileObject(event.getTile(), event.getGameObject(), null);
    }

    @Subscribe
    public void onGroundObjectSpawned(GroundObjectSpawned event)
    {
        onTileObject(event.getTile(), null, event.getGroundObject());
    }

    @Subscribe
    public void onGroundObjectChanged(GroundObjectChanged event)
    {
        onTileObject(event.getTile(), event.getPrevious(), event.getGroundObject());
    }

    @Subscribe
    public void onGroundObjectDespawned(GroundObjectDespawned event)
    {
        onTileObject(event.getTile(), event.getGroundObject(), null);
    }

    @Subscribe
    public void onWallObjectSpawned(WallObjectSpawned event)
    {
        onTileObject(event.getTile(), null, event.getWallObject());
    }

    @Subscribe
    public void onWallObjectChanged(WallObjectChanged event)
    {
        onTileObject(event.getTile(), event.getPrevious(), event.getWallObject());
    }

    @Subscribe
    public void onWallObjectDespawned(WallObjectDespawned event)
    {
        onTileObject(event.getTile(), event.getWallObject(), null);
    }

    @Subscribe
    public void onDecorativeObjectSpawned(DecorativeObjectSpawned event)
    {
        onTileObject(event.getTile(), null, event.getDecorativeObject());
    }

    @Subscribe
    public void onDecorativeObjectChanged(DecorativeObjectChanged event)
    {
        onTileObject(event.getTile(), event.getPrevious(), event.getDecorativeObject());
    }

    @Subscribe
    public void onDecorativeObjectDespawned(DecorativeObjectDespawned event)
    {
        onTileObject(event.getTile(), event.getDecorativeObject(), null);
    }

    @Subscribe
    public void onItemSpawned(ItemSpawned event)
    {
        if (event.getItem().getId() == ItemID.MARK_OF_GRACE)
            marks.add(event.getItem());
    }

    @Subscribe
    public void onItemDespawned(ItemDespawned event)
    {
        if (event.getItem().getId() == ItemID.MARK_OF_GRACE)
            marks.remove(event.getItem());
    }

    private void onTileObject(Tile tile, TileObject oldObject, TileObject newObject)
    {
        obstacles.remove(oldObject);

        if (newObject == null)
        {
            return;
        }

        if (Obstacles.OBSTACLE_IDS.contains(newObject.getId()) ||
                Obstacles.PORTAL_OBSTACLE_IDS.contains(newObject.getId()) ||
                (Obstacles.TRAP_OBSTACLE_IDS.contains(newObject.getId())
                        && Obstacles.TRAP_OBSTACLE_REGIONS.contains(newObject.getWorldLocation().getRegionID())) ||
                Obstacles.SEPULCHRE_OBSTACLE_IDS.contains(newObject.getId()) ||
                Obstacles.SEPULCHRE_SKILL_OBSTACLE_IDS.contains(newObject.getId()))
        {
            obstacles.put(newObject, new Obstacle(tile, null));
        }

        if (Obstacles.SHORTCUT_OBSTACLE_IDS.containsKey(newObject.getId()))
        {
            AgilityShortcut closestShortcut = null;
            int distance = -1;

            // Find the closest shortcut to this object
            for (AgilityShortcut shortcut : Obstacles.SHORTCUT_OBSTACLE_IDS.get(newObject.getId()))
            {
                if (!shortcut.matches(newObject))
                {
                    continue;
                }

                if (shortcut.getWorldLocation() == null)
                {
                    closestShortcut = shortcut;
                    break;
                }
                else
                {
                    int newDistance = shortcut.getWorldLocation().distanceTo2D(newObject.getWorldLocation());
                    if (closestShortcut == null || newDistance < distance)
                    {
                        closestShortcut = shortcut;
                        distance = newDistance;
                    }
                }
            }

            if (closestShortcut != null)
            {
                obstacles.put(newObject, new Obstacle(tile, closestShortcut));
            }
        }
    }

    @Override
    public void paintAboveScene(Graphics2D graphics2D)
    {
        if (marks.size() > 0)
            activeColor = Color.red;
        else
            activeColor = Color.cyan;

        overlayFillColor = new Color(activeColor.darker().getRed(), activeColor.darker().getGreen(), activeColor.darker().getBlue(), 125);

        graphics2D.setColor(activeColor);

        for (TileObject o : obstacles.keySet())
        {
            Color prevColor = activeColor;
            Color prevOverlayFillColor = overlayFillColor;
            AgilityShortcut as = obstacles.get(o).getShortcut();
            if (as != null)
            {
                if (as.getLevel() > client.getBoostedSkillLevel(Skill.AGILITY))
                {
                    activeColor = Color.red;
                }
                else
                {
                    activeColor = Color.green;
                }
                prevOverlayFillColor = overlayFillColor;
                overlayFillColor = new Color(activeColor.darker().getRed(), activeColor.darker().getGreen(), activeColor.darker().getBlue(), 125);
            }
            paintClickBox(graphics2D, o.getPlane(), o.getClickbox());
            activeColor = prevColor;
            overlayFillColor = prevOverlayFillColor;
        }

        for (TileItem tileItem : marks)
        {
            graphics2D.setColor(activeColor);
            graphics2D.draw(Perspective.getCanvasTileAreaPoly(client, tileItem.getTile().getLocalLocation(), 1));
            graphics2D.setColor(Color.white);
            Point p = Perspective.getCanvasTextLocation(client, graphics2D,  tileItem.getTile().getLocalLocation(), MARK_OF_GRACE, OFFSET_Z);
            if (p != null)
                graphics2D.drawString(MARK_OF_GRACE, p.getX(), p.getY());
        }
    }

    private void paintClickBox(Graphics2D graphics2D, int plane, Shape clickbox) {

        if (plane == client.getPlane())
        {
            if (clickbox != null)
            {
                graphics2D.setColor(activeColor);
                graphics2D.draw(clickbox);
                graphics2D.setColor(overlayFillColor);
                graphics2D.fill(clickbox);
            }
        }
    }
}

File: meteor-klient/src/main/kotlin/meteor/plugins/agility/AgilityPlugin.kt
package meteor.plugins.agility

import meteor.Event
import meteor.eventbus.events.*

import meteor.plugins.Plugin
import meteor.plugins.PluginDescriptor
import net.runelite.api.GameState
import net.runelite.api.ItemID
import net.runelite.api.Tile
import net.runelite.api.TileObject
import java.util.ArrayList

@Suppress("UnstableApiUsage")
@PluginDescriptor(
    name = "Agility",
    description = "Show helpful information about agility courses and obstacles",
    enabledByDefault = true,

)
class AgilityPlugin : Plugin() {
    var overlay = overlay<AgilityOverlay>(AgilityOverlay(this))
    val obstacles: MutableMap<TileObject, Obstacle> = HashMap()
    var marks: MutableList<Tile> = ArrayList()

    init {
        registerSubscribers()
    }

    override fun onGameStateChanged(): ((Any) -> Unit) =  { it as GameStateChanged
        if (it.new == GameState.LOADING) {
            marks.clear()
            obstacles.clear()
        }
    }

    override fun onGameObjectSpawned(): ((Any) -> Unit) =  { it as GameObjectSpawned
        onTileObject(it.tile, null, it.gameObject)
    }

    override fun onGameObjectChanged(): ((Any) -> Unit) =  { it as GameObjectChanged
        onTileObject(it.tile, it.oldObject, it.newObject)
    }

    override fun onGameObjectDespawned(): ((Any) -> Unit) =  { it as GameObjectDespawned
        onTileObject(it.tile, it.gameObject, null)
    }

    override fun onGroundObjectSpawned(): ((Any) -> Unit) =  { it as GroundObjectSpawned
        onTileObject(it.tile, null, it.groundObject)
    }

    override fun onGroundObjectChanged(): ((Any) -> Unit) =  { it as GroundObjectChanged
        onTileObject(it.tile, it.previous, it.groundObject)
    }

    override fun onGroundObjectDespawned(): ((Any) -> Unit) =  { it as GroundObjectDespawned
        onTileObject(it.tile, it.groundObject, null)
    }

    override fun onWallObjectSpawned(): ((Any) -> Unit) =  { it as WallObjectSpawned
        onTileObject(it.tile, null, it.wallObject)
    }

    override fun onWallObjectChanged(): ((Any) -> Unit) =  { it as WallObjectChanged
        onTileObject(it.tile, it.previous, it.wallObject)
    }

    override fun onWallObjectDespawned(): ((Any) -> Unit) =  { it as WallObjectDespawned
        onTileObject(it.tile, it.wallObject, null)
    }

    override fun onDecorativeObjectSpawned(): ((Any) -> Unit) =  { it as DecorativeObjectSpawned
        onTileObject(it.tile, null, it.decorativeObject)
    }

    override fun onDecorativeObjectChanged(): ((Any) -> Unit) =  { it as DecorativeObjectChanged
        onTileObject(it.tile, it.previous, it.decorativeObject)
    }

    override fun onDecorativeObjectDespawned(): ((Any) -> Unit) =  { it as DecorativeObjectDespawned
        onTileObject(it.tile, it.decorativeObject, null)
    }

    override fun onItemSpawned(): ((Any) -> Unit) =  { it as ItemSpawned
        if (it.item.id == ItemID.MARK_OF_GRACE) {
            marks.add(it.tile)
        }
    }

    override fun onItemDespawned(): ((Any) -> Unit) =  { it as ItemDespawned
        if (it.item.id == ItemID.MARK_OF_GRACE) {
            marks.remove(it.tile)
        }
    }

    private fun onTileObject(tile: Tile, oldObject: TileObject?, newObject: TileObject?) {
        obstacles.remove(oldObject)
        if (newObject == null) {
            return
        }
        if (Obstacles.OBSTACLE_IDS.contains(newObject.id) ||
            Obstacles.PORTAL_OBSTACLE_IDS.contains(newObject.id) ||
            (Obstacles.TRAP_OBSTACLE_IDS.contains(newObject.id)
                    && Obstacles.TRAP_OBSTACLE_REGIONS.contains(newObject.worldLocation.regionID))
            ||
            Obstacles.SEPULCHRE_OBSTACLE_IDS.contains(newObject.id) ||
            Obstacles.SEPULCHRE_SKILL_OBSTACLE_IDS.contains(newObject.id)
        ) {
            obstacles[newObject] = Obstacle(tile, null)
        }
        if (Obstacles.SHORTCUT_OBSTACLE_IDS!!.containsKey(newObject.id)) {
            var closestShortcut: AgilityShortcut? = null
            var distance = -1

            // Find the closest shortcut to this object
            for (shortcut in Obstacles.SHORTCUT_OBSTACLE_IDS!![newObject.id]) {
                if (!shortcut!!.matches(newObject)) {
                    continue
                }
                if (shortcut.worldLocation == null) {
                    closestShortcut = shortcut
                    break
                } else {
                    val newDistance: Int = shortcut.worldLocation.distanceTo2D(newObject.worldLocation)
                    if (closestShortcut == null || newDistance < distance) {
                        closestShortcut = shortcut
                        distance = newDistance
                    }
                }
            }
            if (closestShortcut != null) {
                obstacles[newObject] = Obstacle(tile, closestShortcut)
            }
        }
    }
}